### PR TITLE
✨[amp-story-player] Adds navigation logic

### LINKED ISSUE #26696: [amp-story-player] Support pre-rendering of next story
Whenever a user is navigating, we should pause the story, play the next one, and pre-render the next-next story; while also reusing a set number of iframes.

### PR DESCRIPTION
**Pull request by Enriqe**
Closes #26696

Tracker #26308

### CODE REVIEW

File: src/amp-story-player.js [hunk 1]
```diff
@@ -248,15 +377,112 @@ export class AmpStoryPlayer {
...
+  /**
+   * Sends a message to the story document to pause it.
+   * @private
+   * @param {number} iframeIdx
+   */
+  pauseStory_(iframeIdx) {
+    this.handshakePromises_[iframeIdx].then(() => {
+      this.messagingFor_[iframeIdx].sendRequest(
+        'visibilitychange',
+        {state: 'paused'},
```
Comment (gmajoulet) [inline src/amp-story-player.js]: To match the other viewers implementations, we should set the document to `inactive`, and keep `paused` for when the swiping gesture is ongoing.

File: src/amp-story-player.js [hunk 2]
```diff
@@ -248,15 +377,112 @@ export class AmpStoryPlayer {
...
 self.onload = () => {
   const manager = new AmpStoryPlayerManager(self);
   manager.loadPlayers();
 };
+
+/**
+ * Manages the iframes used to host the stories inside the player. It keeps
+ * track of the iframes hosting stories, and what stories have an iframe.
+ */
+class IframePool {
```
Comment (gmajoulet) [inline src/amp-story-player.js]: Let's maintain the (now controversial) one class per file rule of thumb?

File: src/amp-story-player.js [hunk 3]
```diff
@@ -193,12 +234,100 @@ export class AmpStoryPlayer {
...
-    this.layoutIframe_(this.stories_[0], this.iframes_[0]);
+    // Display first story.
+    const firstStory = this.stories_[0];
+    const iframeIdx = firstStory[IFRAME_IDX];
+    const storyIframe = this.iframes_[iframeIdx];
+    this.layoutIframe_(firstStory, storyIframe);
+    this.displayStory_(iframeIdx);
+
+    // Pre-render next stories.
+    for (let idx = 1; idx < this.stories_.length && idx < MAX_IFRAMES; idx++) {
```
Comment (gmajoulet) [inline src/amp-story-player.js]: Would be cool to start at `i = 0` and have a parameter in `layoutIframe_` to prerender it or not. This way we could merge this block with the one above.
Eg: `this.layoutIframe_(story, iframe, i !== 0 /** prerender */);` with the `prerender` boolean adding a `visibilitystate=prerender` parameter (cf other comment about prerendering)

File: src/amp-story-player.js [hunk 4]
```diff
@@ -248,15 +377,112 @@ export class AmpStoryPlayer {
...
+
+    /** @private @const {!Array<number>} */
+    this.storyIdsWithIframe_ = [];
+  }
+
+  /**
+   * @param {number} idx
+   * @public
+   */
+  addIframe(idx) {
```
Comment (gmajoulet) [inline src/amp-story-player.js]: Re: `addIframe` accepting a parameter that is not an iframe

File: src/amp-story-player.js [hunk 5]
```diff
@@ -248,15 +377,112 @@ export class AmpStoryPlayer {
...
+    this.iframePool_.push(idx);
+  }
+
+  /**
+   * @param {number} idx
+   * @public
+   */
+  addStory(idx) {
+    this.storyIdsWithIframe_.push(idx);
```
Comment (gmajoulet) [inline src/amp-story-player.js]: Re: naming (ID vs index)

**Comment by Enriqe** [comment]
PTAL

File: src/amp-story-player.js (hunk 2 above)
Comment (gmajoulet) [reply src/amp-story-player.js]: I personally don't feel strongly at all (and am actually probably on your side), but it's the repository's guidelines :man_shrugging:

File: test/unit/test-amp-story-player.js [hunk 6]
```diff
@@ -14,56 +14,126 @@
...
+  it(
+    'should remove iframe from a story with distance > 1 from current story ' +
+      'and give it to a new story that is distance <= 1 when navigating',
+    () => {
+      buildStoryPlayer(4);
+      const stories = toArray(playerEl.querySelectorAll('a'));
+
+      const player = new AmpStoryPlayer(win, playerEl);
+      player.buildCallback();
+      player.layoutCallback();
```
Comment (gmajoulet) [inline test/unit/test-amp-story-player.js]: Can we use `manager.loadPlayers()` here too?
If you need a way to retrieve the `player` instance to call `.next_()` until we can swipe, please add a TODO as well.

File: test/unit/test-amp-story-player.js [hunk 7]
```diff
@@ -14,56 +14,126 @@
...
+
+      // TODO: replace next_() with swipe.
+      player.next_();
+      expect(stories[0]['__AMP_IFRAME_IDX__']).to.eql(0);
```
Comment (gmajoulet) [inline test/unit/test-amp-story-player.js]: Nit: please import the `IFRAME_IDX` constant

**Comment by Enriqe** [comment]
Adding @samouri for owners approval

**Comment by Enriqe** [comment]
PTAL @gmajoulet

File: src/amp-story-player-iframe-pool.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+  /**
+   * Takes the leftmost iframe and allocates it to the next story to the right
+   * without an iframe. It also updates the storyIdsWithIframe by removing the
+   * reference to the detached story and adds the new one.
+   * @param {number} nextStoryIdx
+   * @return {number} Index of the detached story.
+   * @public
+   */
+  rotateLeft(nextStoryIdx) {
```
Comment (gmajoulet) [inline src/amp-story-player-iframe-pool.js]: Nit: please replace left/right with RTL/LTR friendly terms

File: src/amp-story-player.js [hunk 9]
```diff
@@ -193,36 +225,136 @@ export class AmpStoryPlayer {
...
+    if (this.currentIdx_ - 1 < 0) {
+      return;
+    }
+
+    this.currentIdx_--;
+
+    const previousStory = this.stories_[this.currentIdx_ + 1];
+    this.updateVisibilityState_(
+      previousStory[IFRAME_IDX],
+      VisibilityState.PAUSED
```
Comment (gmajoulet) [inline src/amp-story-player.js]: `VisibilityState.INACTIVE`

File: src/amp-story-player.js [hunk 10]
```diff
@@ -193,36 +225,136 @@ export class AmpStoryPlayer {
...
+
+    const nextIframe = this.iframes_[nextStory[IFRAME_IDX]];
+    this.layoutIframe_(nextStory, nextIframe, VisibilityState.PRERENDER);
+    this.setUpMessagingForIframe_(nextStory, nextIframe);
+  }
+
   /**
    * @param {!Element} story
    * @param {!Element} iframe
+   * @param {string} visibilityState
```
Comment (gmajoulet) [inline src/amp-story-player.js]: Nit: here and below, `{!VisibilityState}` types

**Comment by Enriqe** [comment]
Thanks!

**Pull request merged**